--- a/src/arch/s390x.rs
+++ b/src/arch/s390x.rs
@@ -463,16 +463,7 @@ pub unsafe fn on_stack(arg: *mut u8, stack: impl Stack, f: StackCallFunc) {
     // Similar to AArch64, we need to be careful about unwinding information
     // when using .cfi_signal_frame.
     asm_may_unwind_root!(
-        // DW_CFA_GNU_args_size 0
-        cfi_reset_args_size_root!(),
-        concat!("brasl   %r14,", asm_mangle!("stack_call_trampoline")),
-        "nop",                          // NOP for unwinding safety
-        in("r2") arg,
-        in("r3") stack.base().get(),
-        in("r4") f,
-        clobber_abi("C"),
+        "basr    %r14,%r4",             // Call function directly for debugging
+        in("r2") arg,
+        in("r4") f,
     );
 }
